### DISCUSSION TITLE
Mode management

### DISCUSSION THREAD

**Discussion by lukeclements** [nasa/fprime#2310] [Q&A]
Hi, 

How do I implement mode management, specifically on an Arduino microcontroller?

**Comment by LeStarch** [comment]
Can you elaborate on what you mean by mode management?

**Reply by lukeclements** [reply]
The operational modes of the spacecraft during various phases, including nominal operations, data collection, and battery emergency scenarios, among others.

**Reply by LeStarch** [reply]
Mode management is typically done by creating a mode manage component.  It handles the communication required to switch modes (by command or port).  There are two architectures that can be used to communicate mode:

1. `push`, mode manager would have an array of outputs ports that communicate a mode change to components that care. Each component must be responsible for handling that change appropriately.
2. `pull`, mode manager would have a "get current mode port" and each component could call that port to get the current mode and respond appropriately.

@timcanham any other thoughts here?

**Reply by lukeclements** [reply]
That you for the information, is there an example somewhere of this being done?
